### PR TITLE
Ultimate address mode matching

### PR DESCRIPTION
**Pull request by Kha**
Also finally make AddressOperand params RegisterOperands

Another ~8% off of fannkuch

### CODE REVIEW

File: src/main/scala/mjis/RegisterAllocator.scala [hunk 1]
```diff
@@ -451,8 +451,8 @@ class FunctionRegisterAllocator(function: AsmFunction,
         instr.operands(idx) = map(r)
       case (a: AddressOperand, idx) =>
         instr.operands(idx) = a.copy(
-          base = a.base.flatMap(r => Some(map(r.asInstanceOf[RegisterOperand]))),
-          offset = a.offset.flatMap(r => Some(map(r.asInstanceOf[RegisterOperand])))
+          base = a.base.map(map(_).asInstanceOf[RegisterOperand]),
```
Comment (jspam) [inline src/main/scala/mjis/RegisterAllocator.scala]: I don't know who had the terrible idea of naming the parameter `map` \*cough\* but I'm very much in favor of renaming it to `mapping` or similar ...

File: src/main/scala/mjis/AssemblerFileGenerator.scala [hunk 2]
```diff
@@ -38,8 +38,10 @@ class MjisAssemblerFileGenerator(input: AsmProgram, config: Config) extends Asse
     case r: AddressOperand =>
       val params = Seq[Option[String]](
         r.base.map(opToString),
-        r.offset.map(opToString),
-        if (r.scale != 1) Some(r.scale.toString) else None
+        // Semantically, it makes sense for the base to be 8 bytes and the index to be 4 bytes, but that's not how the
+        // assembler syntax likes it
+        r.offset.map(offset => opToString(offset.copy(sizeBytes = r.base.map(_.sizeBytes).getOrElse(offset.sizeBytes)))),
+        if (r.scale != 1 && r.offset.nonEmpty) Some(r.scale.toString) else None
```
Comment (jspam) [inline src/main/scala/mjis/AssemblerFileGenerator.scala]: `if (r.scale != 1) { assert(r.offset.nonEmpty); Some(r.scale.toString) }`?

File: src/main/scala/mjis/opt/FirmExtractors.scala [hunk 3]
```diff
@@ -61,6 +68,14 @@ object FirmExtractors {
     }
   }
 
+  object GenAddExtr {
```
Comment (jspam) [inline src/main/scala/mjis/opt/FirmExtractors.scala]: Ingenious! :tophat:

**Comment by jspam** [comment]
LGTM

File: src/main/scala/mjis/AssemblerFileGenerator.scala (hunk 2 above)
Comment (Kha) [inline src/main/scala/mjis/AssemblerFileGenerator.scala]: Reverted. The assertion is in gcc already, after all :P

**Pull request merged**